### PR TITLE
Settings Dialog Backend

### PR DESCRIPTION
**Pull request by ChowdaaCollege**

### CODE REVIEW

**Comment by iTNTPiston** [comment]
Why is the workflow not running, hmm

**Comment by iTNTPiston** [comment]
I think the workflow is not triggered because your commit is not verified.

**Review by iTNTPiston — CHANGES_REQUESTED** [review]
Please fix the line endings with `just lintpy -f` so the diff is displayed properly

File: packages/celer-web-app/src/core/context/SettingContext.ts [hunk 1]
```diff
@@ -1,22 +1,41 @@
...
-    mapDisplayMode: MapDisplayMode,
-    theme: Theme,
-    splitSetting: SplitTypeSetting<boolean>,
-    enableSubsplits: boolean,
+export type Setting =
+{
+	theme: typeof Themes.Default
+	mapDisplay: typeof MapDisplayModes.Auto
+	splitSettings: typeof defaultSplitSetting
+	enableSubsplits: boolean
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/core/context/SettingContext.ts:11]: Why are you deleting the existing type declaration and using `typeof`? Let's not overuse `typeof` to make the compiler happy. You won't need to change this code if you update the types `MapDisplayMode`, `Theme`, etc to your new system

File: packages/celer-web-app/src/core/context/SettingContext.ts [hunk 2]
```diff
@@ -1,22 +1,41 @@
...
-    setTheme: Consumer<Theme>,
-    setSplitSetting: (value: boolean, ...splitType: SplitType[])=>void,
-    setEnableSubsplits: Consumer<boolean>,
+export const save = (setting: Setting) => { return LocalStorageWrapper.store("Settings", setting);};
+export const load = () => {
+	const stored = LocalStorageWrapper.load("Settings", defaultSettings);
+	const result = {
+		...stored,
+		theme: getNextTheme(stored.theme.name),
+		mapDisplay: getNextMap(stored.mapDisplay.name),
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/core/context/SettingContext.ts:20]: Why are we getting the next theme after loading? This doesn't seem right. If it's intentional please put a comment explaining why

File: packages/celer-web-app/src/core/engine/engine.ts [hunk 3]
```diff
@@ -103,7 +103,7 @@ export class RouteEngine{
 		// this.dupeKorok = [];
 	}
 
-	public setSplitSetting(splitSetting: SplitTypeSetting<boolean>){
+	public setSplitSetting(splitSetting: typeof defaultSplitSetting){
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/core/engine/engine.ts:106]: See my other comment about `typeof`

File: packages/celer-web-app/src/core/settings/MapDisplayMode.ts [hunk 4]
```diff
@@ -32,6 +24,25 @@ export const MapDisplayModes = {
...
+	if(map == "Half")
+	{
+		return MapDisplayModes.Half;
+	}
+	if(map == "Narrow")
+	{
+		return MapDisplayModes.Narrow;
+	}
+	return MapDisplayModes.Hidden;
+};
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/core/settings/MapDisplayMode.ts:48]: Wait.. now I'm really confused. The function is `getNextMap` but what you are doing is essentially `MapDisplayModes[map]` Again if this is intentionally please explain in a comment.

File: packages/celer-web-app/src/core/settings/Theme.ts [hunk 5]
```diff
@@ -12,6 +7,13 @@ export const Themes = {
...
 
-export const ThemeStorage = new SettingStorage("Theme", Themes, Themes.Default);
+export const getNextTheme = (name: string) =>
+{
+	if (name == "Default")
+	{
+		return Themes.Default;
+	}
+	return Themes.Granatus;
+};
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/core/settings/Theme.ts:19]: See my other comment at `getNextMap`

File: packages/celer-web-app/src/ui/styles/StyleEngine.ts [hunk 6]
```diff
@@ -19,7 +19,7 @@ export class StyleEngine<CSArray extends Array<ComputeStyle<StyleMap>>> {
 		this.allStyles = computeStyles;
 	}
 
-	public compute(sizes: AppSizes, colors: AppColors, mapDisplayMode: MapDisplayMode): ComputedStyle<CSArray> {
+	public compute(sizes: AppSizes, colors: AppColors, mapDisplayMode: typeof MapDisplayModes.Auto): ComputedStyle<CSArray> {
 		let mainCss = "";
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/ui/styles/StyleEngine.ts:23]: See my other comment about `typeof`, same below

File: packages/celer-web-app/src/ui/frames/AppFrame.tsx [hunk 7]
```diff
@@ -69,29 +63,76 @@ export const AppFrame: React.FC<EmptyObject> = ()=>{
...
-							<MenuItemWithValue value={getSplitSettingText(splitSetting[SplitType.Shrine])} action={function (): void {
-								setSplitSetting(!splitSetting[SplitType.Shrine], SplitType.Shrine);
+							<MenuItemWithValue value={getSplitSettingText(setting.splitSettings[SplitType.Shrine])} action={function (): void {
+								setSetting({
+									...setting,
+									splitSettings: {
+										...setting.splitSettings,
+										[SplitType.Shrine]: !setting.splitSettings[SplitType.Shrine]
+									}
+								});
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/ui/frames/AppFrame.tsx:73]: Can we use `immer` for manipulating immutable objects? https://immerjs.github.io/immer/

File: packages/celer-web-app/src/ui/frames/AppFrame.tsx [hunk 8]
```diff
@@ -51,7 +54,205 @@ export const AppFrame: React.FC<EmptyObject> = ()=>{
...
+					}} >
+					<div className={styles.menuOverlayFrame}style={showMenu?{ height: "auto" } : undefined}>
+
+						{showMenu && <>
+							{contextMenuRef === splitSettingsMenuItemRef && <div className={styles.submenu} style={{
+								bottom: `calc( 100vh - ${contextMenuRef.current?.getBoundingClientRect().bottom || 0}px )`,
+							}}>
+								<MenuItemWithValue value={getSplitSettingText(setting.splitSettings[SplitType.Shrine])} action={function (): void {
+									setSetting(
+										produce(setting, (draft) => {
```
Comment (iTNTPiston) [inline packages/celer-web-app/src/ui/frames/AppFrame.tsx:78]: You don't need to nest `produce` with `immer`. see https://immerjs.github.io/immer/update-patterns/#nested-data-structures

**Review by iTNTPiston — CHANGES_REQUESTED** [review]

**Comment by iTNTPiston** [comment]
What testing have you done to make sure the features work as expected? (manual/automated?, etc)

**Comment by ChowdaaCollege** [comment]
> What testing have you done to make sure the features work as expected? (manual/automated?, etc)

For each setting, I manually toggled them to make sure both that local storage was being updated after each change and that all the changes were reflected on the screen. Every setting is saved correctly in local storage and is reflected on the screen with both values of the toggle.

**Review by iTNTPiston — APPROVED** [review]

**Pull request merged**